Explain how Z values are asssigned to mesh elements (#7730)
qgis/QGIS-Documentation: docs/user_manual/working_with_mesh/mesh_properties.rst

 +                                       +                         +-------------------------------+------------------------------------------+
 | face or edge of a face                |                         | :guilabel:`Advanced           | :guilabel:`z` widget if in               |
 |                                       |                         | Digitizing Panel` (if         | |locked| :sup:`Locked` state             |
 |                                       |                         | :guilabel:`z` widget is in    |                                          |
 |                                       |                         | |locked| :sup:`Locked` state) |                                          |
 +                                       +-------------------------+-------------------------------+------------------------------------------+
 |                                       | Yes                     | :guilabel:`Vertex Z value`    | Average of the selected vertices         |
 +---------------------------------------+-------------------------+-------------------------------+------------------------------------------+

>  | Vertex on an unshared edge            | No                      | :guilabel:`Vertex Z value`    | Default or user defined                  |
>  +                                       +-------------------------+                               +------------------------------------------+
>  |                                       | Yes                     |                               | Average of the selected vertices         |
>  +---------------------------------------+-------------------------+-------------------------------+------------------------------------------+
>  | Vertex on an edge shared by two faces | ---                     | Mesh layer                    | Interpolated from the faces' vertices    |

Reviewer: ```suggestion
 | Vertex on an edge shared by two faces | ---                     | Mesh layer                    | Interpolated from the edge' vertices    |
```

Author: Any reason to have removed the "vertex on unshared edge" rows?

Reviewer: For vertex on edge inserted on an edge, the behavior is the same if the edge is shared or not by faces: the value is interpolated between the two vertices at the extremity of the edge.
So, better to remove also "shared by two faces" in line 711.

Author: I'm not that sure. See video where numbers represent vertex z value and what happens (new vertex takes value from widget). Peek is bugging here so can't record but if I select vertices before adding the new one, I get the average value assigned.
https://user-images.githubusercontent.com/7983394/193492327-8ff3493a-4e27-4120-836c-59d555bcdde5.mp4

Reviewer: Indeed, may be a bug... I will look at...

 +---------------------------------------+-------------------------+-------------------------------+------------------------------------------+
 | Vertex on a face                      | ---                     | Mesh layer                    | Interpolated from the face vertices      |
 +---------------------------------------+-------------------------+-------------------------------+------------------------------------------+

>  | Vertex snapped to a 2D vector feature | ---                     | ---                           | No value                                 |

Author: Should it be backported?

>  | Vertex snapped to a 2D vector feature | ---                     | ---                           | No value                                 |

Reviewer: The fix should be backported (I just added the label).
If you ask for the doc, maybe it should be backported once the fix will be effectively backported?

>  | Vertex snapped to a 2D vector feature | ---                     | ---                           | No value                                 |

Author: I was asking for code but your reply made me realize that I didn't backport any editing changes to 3.22 docs. So done now at #7818 (allowing you to review the unreviewed bits I merged last week - I'll forward port your suggestions)
